tests/integration/test_target_branch_protection.py

```diff
@@ -0,0 +1,232 @@
+            ["pull request", "reviews", "can be bypassed"],
+            BranchProtectionParameters(
+                branch_protection_enabled=True,
+                require_code_owner_reviews=True,
+                dismiss_stale_reviews_enabled=True,
+                bypass_pull_request_allowance_disabled=False,
+                required_signatures_enabled=False,
+            ),
+            id="pull-request-allowance not empty",
+        ),
+        pytest.param(
+            f"requires-signature/{uuid4()}",
+            ["signed", "commits", "not required"],
+            BranchProtectionParameters(
+                branch_protection_enabled=True,
+                require_code_owner_reviews=True,
+                dismiss_stale_reviews_enabled=True,
+                bypass_pull_request_allowance_disabled=True,
+                required_signatures_enabled=False,
+            ),
+            id="required-signature disabled",
+        ),
+    ],
+    indirect=["github_branch"],
+)
+def test_fail(
+    github_branch: Branch,
+    reason_string_array: str,
+    branch_protection_parameters: BranchProtectionParameters,
+    github_repository_name: str,
+):
+    """
+    arrange: given a branch that is not compliant.
+    act: when target_branch_protection is called with the name of the branch.
+    assert: then a fail report is returned.
+    """
+    if branch_protection_parameters.branch_protection_enabled:
+        edit_branch_protection(github_branch, branch_protection_parameters)
+
+    # The github_client is injected
+    report = target_branch_protection(  # pylint: disable=no-value-for-parameter
+        repository_name=github_repository_name, branch_name=github_branch.name
+    )
+
+    if branch_protection_parameters.branch_protection_enabled:
+        github_branch.remove_protection()
+
+    assert_substrings_in_string(reason_string_array, str(report.reason))
+    assert github_branch.name in str(report.reason)
+    assert report.result == Result.FAIL
+
+
+@pytest.mark.parametrize("github_branch", [f"protected/{uuid4()}"], indirect=True)
+def test_pass(
+    github_branch: Branch,
+    github_repository_name: str,
+):
+    """
+    arrange: given a branch that is compliant.
+    act: when target_branch_protection is called with the name of the branch.
+    assert: then a pass report is returned.
+    """
+    branch_protection_parameters = BranchProtectionParameters(
+        branch_protection_enabled=True,
+        require_code_owner_reviews=True,
+        dismiss_stale_reviews_enabled=True,
+        bypass_pull_request_allowance_disabled=True,
+        required_signatures_enabled=True,
+    )
+    edit_branch_protection(github_branch, branch_protection_parameters)
+
+    # The github_client is injected
+    report = target_branch_protection(  # pylint: disable=no-value-for-parameter
+        repository_name=github_repository_name, branch_name=github_branch.name
+    )
+    github_branch.remove_protection()
```

Comment:
Should this also be in a fixture for cleanup?

Reply:
Done